src/Avalonia.Base/Input/GestureRecognizers/VelocityTracker.cs

```diff
@@ -0,0 +1,402 @@
+{
+    // TODO: add 'IOSScrollViewFlingVelocityTracker' and 'MacOSScrollViewFlingVelocityTracker'?
+
+    public readonly record struct Velocity(Vector PixelsPerSecond)
+    {
+        public Velocity ClampMagnitude(double minValue, double maxValue)
+        {
+            Debug.Assert(minValue >= 0.0);
+            Debug.Assert(maxValue >= 0.0 && maxValue >= minValue);
+            double valueSquared = PixelsPerSecond.SquaredLength;
+            if (valueSquared > maxValue * maxValue)
+            {
+                double length = PixelsPerSecond.Length;
+                return new Velocity(length != 0.0 ? (PixelsPerSecond / length) * maxValue : Vector.Zero);
+                // preventing double.NaN in Vector PixelsPerSecond is important -- if a NaN eventually gets into a
+                // ScrollGestureEventArgs it results in runtime errors.
+            }
+            if (valueSquared < minValue * minValue)
+            {
+                double length = PixelsPerSecond.Length;
+                return new Velocity(length != 0.0 ? (PixelsPerSecond / length) * minValue : Vector.Zero);
+            }
+            return this;
+        }
+    }
+
+    /// A two dimensional velocity estimate.
+    ///
+    /// VelocityEstimates are computed by [VelocityTracker.getVelocityEstimate]. An
+    /// estimate's [confidence] measures how well the velocity tracker's position
+    /// data fit a straight line, [duration] is the time that elapsed between the
+    /// first and last position sample used to compute the velocity, and [offset]
+    /// is similarly the difference between the first and last positions.
+    ///
+    /// See also:
+    ///
+    ///  * [VelocityTracker], which computes [VelocityEstimate]s.
+    ///  * [Velocity], which encapsulates (just) a velocity vector and provides some
+    ///    useful velocity operations.
+    public record VelocityEstimate(Vector PixelsPerSecond, double Confidence, TimeSpan Duration, Vector Offset);
+
+    internal record struct PointAtTime(bool Valid, Vector Point, TimeSpan Time);
+
+    /// Computes a pointer's velocity based on data from [PointerMoveEvent]s.
+    ///
+    /// The input data is provided by calling [addPosition]. Adding data is cheap.
+    ///
+    /// To obtain a velocity, call [getVelocity] or [getVelocityEstimate]. This will
+    /// compute the velocity based on the data added so far. Only call these when
+    /// you need to use the velocity, as they are comparatively expensive.
+    ///
+    /// The quality of the velocity estimation will be better if more data points
+    /// have been received.
+    public class VelocityTracker
+    {
+        private const int AssumePointerMoveStoppedMilliseconds = 40;
+        private const int HistorySize = 20;
+        private const int HorizonMilliseconds = 100;
+        private const int MinSampleSize = 3;
+        private const double MinFlingVelocity = 50.0; // Logical pixels / second (defined in flutter\lib\src\gesture\constants.dart)
```

Comment:
Dart related part of the comment can be removed, as we don't have these constants.